File: frontend/src/api/masterData.ts
import { apiRequest } from "./auth";
import type { PageResult, TestStage } from "./stageData";

export type ProductCrosswalkStatus = "PENDING" | "APPROVED" | "REJECTED" | "RETIRED";

export interface ProductCrosswalk {
  crosswalk_id: number;
  supplier_id: number;
  supplier_code: string;
  supplier_name: string;
  test_stage: TestStage;
  raw_product_code: string;
  product_id: number;
  tms_product_code: string;
  identity_class: string;
  enterprise_system: string;
  enterprise_key: string | null;
  status: ProductCrosswalkStatus;
  first_observed_at_utc: string;
  last_observed_at_utc: string;
  approved_by_login: string | null;
  approved_at_utc: string | null;
  decision_reason: string | null;
}

export interface ProductCrosswalkRequest {
  page: number;
  page_size: number;
  status?: ProductCrosswalkStatus;
  supplier_code?: string;
  test_stage?: TestStage;
  raw_product_code?: string;
}

export interface ApproveProductCrosswalkPayload {
  enterprise_system: "SAP_B1";
  enterprise_key: string;
  reason: string;
}

export const listProductCrosswalks = (request: ProductCrosswalkRequest) => {
  const query = new URLSearchParams({ page: String(request.page), page_size: String(request.page_size) });
  for (const key of ["status", "supplier_code", "test_stage", "raw_product_code"] as const) {
    const value = request[key]?.trim();
    if (value) query.set(key, value);
  }
  return apiRequest<PageResult<ProductCrosswalk>>(`/api/v1/master-data/product-crosswalks?${query}`);
};

export const approveProductCrosswalk = (crosswalkId: number, payload: ApproveProductCrosswalkPayload) =>
  apiRequest<ProductCrosswalk>(`/api/v1/master-data/product-crosswalks/${crosswalkId}/approve`, {
    method: "POST",
    body: JSON.stringify(payload),
  });

export const rejectProductCrosswalk = (crosswalkId: number, reason: string) =>
  apiRequest<ProductCrosswalk>(`/api/v1/master-data/product-crosswalks/${crosswalkId}/reject`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  });

File: backend/app/domain/master_data.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.domain.auth import Principal


@dataclass(frozen=True, slots=True)
class ProductCrosswalk:
    crosswalk_id: int
    supplier_id: int
    supplier_code: str
    supplier_name: str
    test_stage: str
    raw_product_code: str
    product_id: int
    tms_product_code: str
    identity_class: str
    enterprise_system: str
    enterprise_key: str | None
    status: str
    first_observed_at_utc: str
    last_observed_at_utc: str
    approved_by_login: str | None
    approved_at_utc: str | None
    decision_reason: str | None


@dataclass(frozen=True, slots=True)
class ProductCrosswalkPage:
    items: tuple[ProductCrosswalk, ...]
    total: int
    page: int
    page_size: int


class ApproveProductCrosswalkRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    enterprise_system: str = Field(default="SAP_B1", pattern=r"^SAP_B1$")
    enterprise_key: str = Field(min_length=1, max_length=128)
    reason: str = Field(min_length=4, max_length=1000)

    @field_validator("enterprise_key")
    @classmethod
    def validate_key(cls, value: str) -> str:
        if any(ord(character) < 32 for character in value):
            raise ValueError("enterprise_key contains control characters")
        return value


class RejectProductCrosswalkRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    reason: str = Field(min_length=4, max_length=1000)


class MasterDataService(Protocol):
    def list_product_crosswalks(
        self,
        *,
        page: int,
        page_size: int,
        status: str | None = None,
        supplier_code: str | None = None,
        test_stage: str | None = None,
        raw_product_code: str | None = None,
    ) -> ProductCrosswalkPage: ...

    def approve_product_crosswalk(
        self,
        crosswalk_id: int,
        request: ApproveProductCrosswalkRequest,
        principal: Principal,
    ) -> ProductCrosswalk: ...

    def reject_product_crosswalk(
        self,
        crosswalk_id: int,
        request: RejectProductCrosswalkRequest,
        principal: Principal,
    ) -> ProductCrosswalk: ...
